### PR TITLE
Accounts Aging report API

### PR DESCRIPTION
**Pull request by mohinid**
## Notion card
https://www.notion.so/saeloun/Backend-for-Accounts-Aging-report-4ad19cb2f9cb4204a356e533757a7f31

## Summary
The accounts Aging report should be available to the owner or admin to check which of their clients are taking longer time settle their invoices. 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Local, RSPec, Postman

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [x] I have added automated tests for my code

### CODE REVIEW

**Comment by mohinid** [comment]
**JSON Response-** 
```

{
    "report": {
        "clients": [
            {
                "id": 2,
                "name": "Flipkart",
                "amount_overdue": {
                    "zero_to_thirty_days": 0,
                    "thirty_one_to_sixty_days": 0,
                    "sixty_one_to_ninety_days": 0,
                    "ninety_plus_days": "2000.0",
                    "total": "2000.0"
                }
            }
        ],
        "total_amount_overdue": {
            "zero_to_thirty_days": 0,
            "thirty_one_to_sixty_days": 0,
            "sixty_one_to_ninety_days": 0,
            "ninety_plus_days": "2000.0",
            "grand_total": "2000.0"
        }
    }
}
```

cc @akhilgkrishnan

**Comment by mohinid** [comment]
Adding search for inclusive dates. Working fine in query.
For eg today is 17th Nov then the result would be with dates -
	```
amount_overdue: {
				“0-30 days”: , =17 nov - 17 oct
				“31-60 days”: , =17 sept- 16 oct(31 day)
				“61-90 days”: , =17 aug - 16 sept
				”90+ days”: , 17 year back - 16 aug
				“total”:
			}
```
Screenshot 2022-11-17 at 10 36 36 AM

**Comment by mohinid** [comment]
@apoorv1316 the cypress tests are failing on this pr, can you please check.

File: Gemfile.lock [hunk 1]
```diff
@@ -545,7 +545,6 @@ DEPENDENCIES
   foreman
   grover
   hash_dot
-  honeybadger
```
Comment (vipulnsward) [inline Gemfile.lock:548]: Please revert

File: app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+json.report do
+  json.clients clients do |client|
+    json.id client.id
+    json.name client.name
+    json.amount_overdue do
+      json.zero_to_thirty_days client.invoices.overdue
+        .during((1.month.ago)...(Date.today + 1.day)).pluck(:amount_due).sum
+      json.thirty_one_to_sixty_days client.invoices.overdue
```
Comment (vipulnsward) [inline app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder:10]: So many queries in view. This should all be done via View Presenter or service. Please search blog on view services

File: config/routes/internal_api.rb [hunk 3]
```diff
@@ -23,8 +23,8 @@
           get :download
         end
       end
-
```
Comment (vipulnsward) [inline config/routes/internal_api.rb:26]: Revert if unrelated

**Review by vipulnsward — COMMENTED** [review]
@mohinid please modify all views to remove queries and use view services/presenters

File: app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder (hunk 2 above)
Comment (mohinid) [reply app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder:10]: Yea, need to refactor.. Checking!

File: config/routes/internal_api.rb (hunk 3 above)
Comment (mohinid) [reply config/routes/internal_api.rb:26]: No, here I added the new route for accounts_aging. I think it's ok.

File: Gemfile.lock (hunk 1 above)
Comment (mohinid) [reply Gemfile.lock]: done

**Comment by apoorv1316** [comment]
@mohinid Can you please share the postman request screenshot along with all the required params and other info.

**Comment by mohinid** [comment]
> @mohinid Can you please share the postman request screenshot along with all the required params and other info.

Sure.

File: app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+  json.total_amount_overdue do
+    json.zero_to_thirty_days invoice_overdue
+      .during((1.month.ago)...(Date.today + 1.day)).pluck(:amount_due).sum
+    json.thirty_one_to_sixty_days invoice_overdue
+      .during((2.month.ago)...1.month.ago).pluck(:amount_due).sum
+    json.sixty_one_to_ninety_days invoice_overdue
+      .during((3.month.ago)...2.month.ago).pluck(:amount_due).sum
+    json.ninety_plus_days invoice_overdue
```
Comment (vipulnsward) [inline app/views/internal_api/v1/reports/accounts_aging/index.json.jbuilder:26]: ` json.ninety_plus_days @aging_presenter. ninety_plus_days`

**Comment by mohinid** [comment]
> @mohinid please modify all views to remove queries and use view services/presenters

Hi Vipul, did some refactoring. please review now.

**Comment by mohinid** [comment]
> @mohinid Can you please share the postman request screenshot along with all the required params and other info.

@apoorv1316 you need to hit this url- **http://localhost:3000/internal_api/v1/reports/accounts_aging** with a **get** request.
You should be able to see a list of all clients in the current company(check what the current workspace ID of the user is) and their overdue amount based on the date range, In the end, you should be able to see the total sum of overdue invoices in date range as well as grand total as shown in the response below.
Screenshot 2022-11-29 at 4 20 15 PM

Let me know if you need more info. 
**Please test for all scenarios by creating invoices in the different date ranges. You can set due_date- some past date and status-overdue from DB.**

**Comment by mohinid** [comment]
> @mohinid please see name changes on https://github.com/saeloun/miru-web/pull/781/commits/d946bc46885d41302e0b16571a5ee701f9b8cf31
> 
> 
> 
> The code should be easier to read
> 
> 
> 
> Rebase and merge PR on green

Thanks!

**Comment by mohinid** [comment]
@apoorv1316 please approve and merge today. All changes are implemented.

**Review by apoorv1316 — APPROVED** [review]

**Pull request merged**